distribution/lib/Standard/Test/0.0.0-dev/src/Extensions.enso

```diff
@@ -319,12 +322,17 @@ Error.should_be_false self = Test.fail_match_on_unexpected_error self 1
 
          example_should_be_a = 1.should_be_a Boolean
 Any.should_be_a : Any -> Test_Result
-Any.should_be_a self typ = if self.is_a typ || self==typ then Test_Result.Success else
-    loc = Meta.get_source_location 0
-    expected_type = Meta.get_qualified_type_name typ
-    actual_type = Meta.get_qualified_type_name self
-    message = "Expected a value of type " + expected_type + " but got a value of type " + actual_type + " instead (at " + loc + ")."
-    Test.fail message
+Any.should_be_a self typ =
+    ok = case Meta.meta typ of
+        c : Meta.Constructor -> Meta.meta self . constructor == c
+        _ -> self.is_a typ || self==typ
+
+    if ok then Test_Result.Success else
+        loc = Meta.get_source_location 0
+        expected_type = Meta.get_qualified_type_name typ
+        actual_type = Meta.get_qualified_type_name self
+        message = "Expected a value of type " + expected_type + " but got a value of type " + actual_type + " instead (at " + loc + ")."
+        Test.fail message
```

Comment:
```suggestion
Any.should_be_a self typ =
    loc = Meta.get_source_location 0
    case Meta.meta typ of
        c : Meta.Constructor ->
        	ok = case Meta.meta self of
        	    a : Meta.Atom -> a.constructor == c
        	    _ -> False
        	if ok then Test_Result.Success else
		        expected_ctor = Meta.get_qualified_type_name typ + "." + c.name
		        actual_ctor = Meta.get_qualified_type_name self + "." + a.constructor.name
		        message = "Expected a value matching constructor " + expected_ctor+ " but got a value that was constructed with " + actual_ctor + " instead (at " + loc + ")."
		        Test.fail message
        _ -> 
        	ok = self.is_a typ || self==typ

		    if ok then Test_Result.Success else
		        expected_type = Meta.get_qualified_type_name typ
		        actual_type = Meta.get_qualified_type_name self
		        message = "Expected a value of type " + expected_type + " but got a value of type " + actual_type + " instead (at " + loc + ")."
		        Test.fail message
```


If I do `(File_Error.Access_Denied file).should_be_a File_Error.Not_Found`, I want the error message to be clear.

Note that `Meta.get_qualified_type_name (File_Error.Not_Found) == "[...].File_Error"`.

With the current solution, I could get `Expected a value of type [...].File_Error but got a value of type [...].File_Error`. Uh that will be confusing!

I think we need to handle the constructor case as a completely separate case to display a good error message.

Reply:
Radek, feel free to push your changes into my branch.